nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py

```diff
@@ -350,93 +353,63 @@ def finish_validation_step(self):
             self.cfg.get("reset_position_ids", False),
             self.cfg.get("reset_attention_mask", False),
             self.cfg.get("eod_mask_loss", False),
         )
-        micro_batch_size = context_tokens_tensor.shape[0]
-        sequence_length = context_tokens_tensor.shape[1]
-
-        app_state = AppState()
-        _reconfigure_microbatch_calculator(
-            rank=app_state.global_rank,
-            rampup_batch_size=None,
-            global_batch_size=micro_batch_size,
-            micro_batch_size=micro_batch_size,
-            data_parallel_size=1,
-        )
-        attention_mask_repeat = torch.concat([attention_mask for _ in range(micro_batch_size)])
-        rewards = self.forward_step(
-            [context_tokens_tensor, context_length_tensor, position_ids, attention_mask_repeat],
-            micro_batch_size,
-            sequence_length,
-        )
+        attention_mask = attention_mask.expand(inference_batch_size, -1, -1, -1)
+        inputs = [context_tokens_tensor, context_length_tensor, position_ids, attention_mask]
+
+        # if it's not divisible by forward micro batch size,
+        # then try our best to find a smaller one that is divisible
+        while inference_batch_size % forward_micro_batch_size != 0:
+            forward_micro_batch_size -= 1
+
+        num_microbatches = divide(inference_batch_size, forward_micro_batch_size)
+        data_iter = get_iterator_k_split(inputs, num_microbatches)
+
+        rewards = self.forward_step(data_iter, forward_micro_batch_size, sequence_length, num_microbatches)
 
         if parallel_state.is_pipeline_last_stage():
-            rewards = rewards[0]["reward"]
+            rewards = torch.cat(rewards)
 
             # Standardize values to subtract a bias.
             if self.enable_standardization:
                 rewards = (rewards - self.rew_mean) / self.rew_std
 
         rewards = broadcast_2d_tensor_within_pp(rewards)
+        return rewards
 
-        rewards_list = gather_tensor(
-            rewards, dst=parallel_state.get_data_parallel_src_rank(), group=parallel_state.get_data_parallel_group()
-        )
-
-        return rewards_list, exceeded
-
-    def forward_step(self, batch, micro_batch_size, sequence_length):
+    def forward_step(self, data_iter, micro_batch_size, sequence_length, num_microbatches):
         set_sync_funcs(self, forward_only=True)
 
         fwd_bwd_function = get_forward_backward_func()
         output_tensor = fwd_bwd_function(
             forward_step_func=self.get_forward_output_only_func(),
-            data_iterator=iter([batch]),
+            data_iterator=self._make_data_iterator_list(data_iter),
             model=self.model,
-            num_microbatches=get_num_microbatches(),
+            num_microbatches=num_microbatches,
             forward_only=True,
             seq_length=sequence_length,
             micro_batch_size=micro_batch_size,
+            collect_non_loss_data=True,
```

Comment:
there is a bug in the critic where depending on the number of inference microbatches, the values will be scaled down.

This is because by default megatron-core thinks we're in training mode and therefore will scale our values by num_microbatches [here](https://github.com/NVIDIA/Megatron-LM/blob/c7a1f82d761577e6ca0338d3521eac82f2aa0904/megatron/core/pipeline_parallel/schedules.py#L231). We need to set `collect_non_loss_data=True` to disable this behaviour so it doesn't do the num micro batches scaling when we run inference https://github.com/NVIDIA/Megatron-LM/blob/c7a1f82d761577e6ca0338d3521eac82f2aa0904/megatron/core/pipeline_parallel/schedules.py#L239